### PR TITLE
fix flaky array_item test failures

### PR DESCRIPTION
**Pull request by binmahone**
This PR closes https://github.com/NVIDIA/spark-rapids/issues/8652

Currently, the cache key for ArrayGen (as well as some other Gens) is incomplete, which will lead to unexpected input being reused.
For the test case described in the issue, it's veriying GetArrayElement(array, index) will throw an exception when index is a invalid value(like negative values). However, if this test case is preceded by another test case which will populate cache with an array with all nulls, GetArrayElement will return null regardless of the index value (Thus not throwing exceptions).

The solution is to refine cache key for all of the XXGen.

### CODE REVIEW

**Comment by binmahone** [comment]
build

File: integration_tests/src/main/python/parquet_write_test.py [hunk 1]
```diff
@@ -224,7 +224,11 @@ def test_all_null_int96(spark_tmp_path):
     class AllNullTimestampGen(TimestampGen):
         def start(self, rand):
             self._start(rand, lambda : None)
-    data_path = spark_tmp_path + '/PARQUET_DATA'
+
+        def _cache_repr(self):
+            return super()._cache_repr() + '(all_nulls)'
+
+data_path = spark_tmp_path + '/PARQUET_DATA'
```
Comment (thirtiseven) [inline integration_tests/src/main/python/parquet_write_test.py:231]: Is the indentation here correct?

**Comment by binmahone** [comment]
build

File: integration_tests/src/main/python/parquet_write_test.py [hunk 2]
```diff
@@ -224,7 +224,11 @@ def test_all_null_int96(spark_tmp_path):
     class AllNullTimestampGen(TimestampGen):
         def start(self, rand):
             self._start(rand, lambda : None)
-    data_path = spark_tmp_path + '/PARQUET_DATA'
+
+        def _cache_repr(self):
```
Comment (thirtiseven) [inline integration_tests/src/main/python/parquet_write_test.py:228]: So we need to add a `_cache_repr` even if it will only be used for once to prevent a key conflict in cache. Didn't realized it, nice catch.

File: integration_tests/src/main/python/data_gen.py [hunk 3]
```diff
@@ -769,12 +774,12 @@ def __init__(self, min_value=MIN_DAY_TIME_INTERVAL, max_value=MAX_DAY_TIME_INTER
...
-        start_index = fields.index(start_field)
-        end_index = fields.index(end_field)
-        if start_index > end_index:
+        self._start_index = fields.index(start_field)
+        self._end_index = fields.index(end_field)
+        if  self._start_index > self._end_index:
             raise RuntimeError('Start field {}, end field {}, valid fields is {}, start field index should <= end '
                                'field index'.format(start_field, end_field, fields))
-        super().__init__(DayTimeIntervalType(start_index, end_index), nullable=nullable, special_cases=special_cases)
+        super().__init__(DayTimeIntervalType( self._start_index, self._end_index), nullable=nullable, special_cases=special_cases)
```
Comment (thirtiseven) [inline integration_tests/src/main/python/data_gen.py:782]: nit: an extra white space?

File: integration_tests/src/main/python/data_gen.py [hunk 4]
```diff
@@ -769,12 +774,12 @@ def __init__(self, min_value=MIN_DAY_TIME_INTERVAL, max_value=MAX_DAY_TIME_INTER
         self._min_micros = (math.floor(min_value.total_seconds()) * 1000000) + min_value.microseconds
         self._max_micros = (math.floor(max_value.total_seconds()) * 1000000) + max_value.microseconds
         fields = ["day", "hour", "minute", "second"]
-        start_index = fields.index(start_field)
-        end_index = fields.index(end_field)
-        if start_index > end_index:
+        self._start_index = fields.index(start_field)
+        self._end_index = fields.index(end_field)
+        if  self._start_index > self._end_index:
```
Comment (thirtiseven) [inline integration_tests/src/main/python/data_gen.py:779]: nit: an extra white space?

**Review by thirtiseven — APPROVED** [review]
The array integration test issue is quite confusing to me even though it comes from some code I wrote. Thanks for investigating and fixing it!

**Review by revans2 — APPROVED** [review]

**Review by jlowe — APPROVED** [review]
Excellent find, @binmahone!  🚀
